dartagnan/src/main/java/com/dat3m/dartagnan/program/analysis/BackwardsReachingDefinitionsAnalysis.java

```diff
@@ -0,0 +1,384 @@
+        }
+    }
+
+    private static final class RegisterWritersView implements RegisterWriters {
+
+        private final ReaderInfo info;
+        private final Register register;
+
+        private RegisterWritersView(ReaderInfo info, Register register) {
+            this.info = info;
+            this.register = register;
+        }
+
+        @Override
+        public List<RegWriter> getMayWriters() {
+            return info.mayWriters.stream().filter(w -> w.getResultRegister().equals(register)).toList();
+        }
+
+        @Override
+        public List<RegWriter> getMustWriters() {
+            return info.mustWriters.stream().filter(w -> w.getResultRegister().equals(register)).toList();
+        }
+
+        @Override
+        public boolean mustBeInitialized() {
+            return !info.uninitialized.contains(register);
+        }
+    }
+
+    /**
+     * Local information for either a {@link RegWriter} or the initial state of the program.
+     */
+    public static final class Readers {
+
+        private final Set<RegReader> readers = new HashSet<>();
+        private boolean mayBeFinal;
+
+        private Readers() {}
+
+        /**
+         * Lists readers s.t. there may be an execution where an instance of that reader uses the result of an instance of the associated writer.
+         * @return Unordered set of Readers.
+         */
+        public Collection<RegReader> getReaders() {
+            return Collections.unmodifiableCollection(readers);
+        }
+
+        /**
+         * @return {@code true} if there may be an execution, where an instance of the associated writer defines the last value of its register.
+         */
+        public boolean mayBeFinal() {
+            return mayBeFinal;
+        }
+    }
+
+    private BackwardsReachingDefinitionsAnalysis() {}
+
+    private void initialize(Function function, Set<Register> finalRegisters) {
+        for (Event event : function.getEvents()) {
+            if (event instanceof RegWriter writer) {
+                writerMap.put(writer, new Readers());
+            }
+        }
+        writerMap.put(null, new Readers());
+        for (Event event : function.getEvents()) {
+            if (event instanceof RegReader reader) {
+                final Set<Register> usedRegisters = new HashSet<>();
+                for (Register.Read read : reader.getRegisterReads()) {
+                    usedRegisters.add(read.register());
+                }
+                readerMap.put(reader, new ReaderInfo(usedRegisters));
+            }
+        }
+        readerMap.put(null, new ReaderInfo(finalRegisters));
+    }
+
+    private void run(Function function, Set<Register> finalRegisters) {
+        //For each register used after this state, all future users.
+        final State currentState = new State();
+        for (Register finalRegister : finalRegisters) {
+            final var info = new StateInfo();
+            info.mayReaders.add(null);
```

Comment:
You are abusing the usage of `null` for some special purpose that is not clear to me at all. Is this really necessary?